### PR TITLE
Allocate a matrix in the rastest on the heap

### PR DESCRIPTION
**Pull request by samolisov**
A matrix in the method `static void matrixSquare(void)` is allocated on
the stack:

```c
uintptr_t matrix[MATRIX_SIZE][MATRIX_SIZE]
```

Because the constant `MATRIX_SIZE` is defined as `1000`, the amount of
memory needed to allocate `matrix` equals to `8000000`. This value is more
than the default stack size for the major C/C++ toolchains. In order to
protect the test from the Stack Overflow error, the `matrix` should be
allocated on the heap as discussed in #2386.

Closes: #2386
Signed-off-by: Pavel Samolysov <samolisov@gmail.com>

### CODE REVIEW

File: fvtest/rastest/cpuLoadAgent.c [hunk 1]
```diff
@@ -401,15 +402,17 @@ matrixSquare(void)
 {
 	/* Declare and initialize the matrix. memset works here since this is a true 2D
 	 * array defined in the same scope as the memset. */
-	uintptr_t matrix[MATRIX_SIZE][MATRIX_SIZE];
+	uint32_t size_of_matrix = MATRIX_SIZE * MATRIX_SIZE * sizeof(uintptr_t);
+	uintptr_t *matrix = (uintptr_t*) malloc(size_of_matrix);
 	uintptr_t i = 0;
 	uintptr_t j = 0;
-
```
Comment (charliegracie) [inline fvtest/rastest/cpuLoadAgent.c]: Now that we are using malloc to allocate the memory can we add a NULL check on the result and then just print an error message and return if we fail to allocate?

Comment (charliegracie) [reply fvtest/rastest/cpuLoadAgent.c]: Both of the callers have an OMR_VMTHread so you can pass that through as the parameter.  Once you have the thread you can insert the following macro to get access to the PortLibrary for printing `OMRPORT_ACCESS_FROM_OMRVMTHREAD(vmThread);`

Now that you have access to the PortLibrary you can use the `omrtty_printf` function to write information in the case where we fail to malloc memory.

**Comment by samolisov** [comment]
*@charliegracie* What is your opinion about whether this method is required
at all? Please see the discussion here #2386 If the method is required,
I'll implement your suggestions.

File: fvtest/rastest/cpuLoadAgent.c (hunk 1 above)
Comment (charliegracie) [reply fvtest/rastest/cpuLoadAgent.c]: oh and also you should be using the port library `omrmem_allocate_memory` and `omrmem_free_memory` APIs

**Comment by charliegracie** [comment]
I would keep the method for now.

**Comment by samolisov** [comment]
@charliegracie thank you for the comments. I've applied what you suggested. One moment I don't understand clearly is memory categories, I use the `OMRMEM_CATEGORY_UNKNOWN` one:

```c
uintptr_t *matrix = (uintptr_t*) omrmem_allocate_memory(size_of_matrix,
			OMRMEM_CATEGORY_UNKNOWN);
```
Is it OK?

File: fvtest/rastest/cpuLoadAgent.c (hunk 1 above)
Comment (rwy0717) [reply fvtest/rastest/cpuLoadAgent.c]: I would change `MATRIX_SIZE` to `MATRIX_ROWS` (or something) and use the symbol `MATRIX_SIZE` to represent `size_of_matrix`.

**Comment by samolisov** [comment]
I can change MATRIX_SIZE to whatever but MATRIX_ROW says only about a
number of rows but what about columns?

Wikipedia says:  An *n*-by-*n* matrix is known as a square matrix of order
*n*

*What about MATRIX_ORDER?*

2018-03-28 14:33 GMT+03:00 Robert Young <notifications@github.com>:

**Comment by rwy0717** [comment]
This is not the first time we've had issues with CPU burners being optimized away. Maybe we should move the burn function to a separate compilation unit (EG rasTestHelpers.cpp), so the work is less likely to evaporate. We also have other burners in processTimeTest, we should really only have one common implementation, given how finicky they are to implement.

I prefer processTimeTest's implementation, since the atomic ops are less likely to be optimized away.

```c++
void
userTimeCpuBurn()
{
	for (size_t i = 0; i < 40000; i++) {
		VM_AtomicSupport::nop();
	}
}
```

And sure, matrix order sounds fine to me, if we keep the current implementation.

**Comment by charliegracie** [comment]
Do not worry about the Atomic changes.  I think that would be best in another PR.  Thanks for the great work @samolisov

**Comment by charliegracie** [comment]
@genie-omr build all

**Comment by charliegracie** [comment]
Actually I missed one thing.  I missed the copyright headers, which need to be updated to 2018.  Can you push a new commit which updates the 2 copyright headers?  You can put `[ci skip]` in the commit title to skip all CI testing since it is just a documentation change.

**Comment by samolisov** [comment]
@charliegracie Done!

**Pull request merged**